sshutil/fingerprint.go

```diff
@@ -0,0 +1,150 @@
+package sshutil
+
+import (
+	"crypto/dsa" //nolint:staticcheck // support for DSA fingerprints
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"go.step.sm/crypto/internal/emoji"
+	"golang.org/x/crypto/ssh"
+)
+
+// FingerprintEncoding defines the supported encodings in SSH key and
+// certificate fingerprints.
+type FingerprintEncoding int
+
+// Supported fingerprint encodings.
+const (
+	// Base64RawFingerprint represents the base64RawStd encoding of the
+	// fingerprint. This is the default encoding for an SSH key.
+	Base64RawFingerprint FingerprintEncoding = iota
+	// Base64RawURLFingerprint represents the base64RawURL encoding of the
+	// fingerprint.
+	Base64RawURLFingerprint
+	// Base64Fingerprint represents the base64 encoding of the fingerprint.
+	Base64Fingerprint
+	// Base64UrlFingerprint represents the base64URL encoding of the
+	// fingerprint.
+	Base64UrlFingerprint FingerprintEncoding = iota
+	// HexFingerprint represents the hex encoding of the fingerprint.
+	HexFingerprint
+	// EmojiFingerprint represents the emoji encoding of the fingerprint.
+	EmojiFingerprint
+)
+
+// Fingerprint returns the SHA-256 fingerprint of an ssh public key or
+// certificate.
+func Fingerprint(pub ssh.PublicKey) string {
+	return EncodedFingerprint(pub, Base64RawFingerprint)
+}
+
+// EncodedFingerprint returns an encoded the SHA-256 fingerprint of an ssh
+// public key or certificate using the specified encoding. In an invalid
+// encoding is passed, the return value will be an empty string.
+func EncodedFingerprint(pub ssh.PublicKey, encoding FingerprintEncoding) string {
+	const prefix = "SHA256:"
+
+	sum := sha256.Sum256(pub.Marshal())
+	switch encoding {
+	case Base64RawFingerprint:
+		return prefix + base64.RawStdEncoding.EncodeToString(sum[:])
+	case Base64RawURLFingerprint:
+		return prefix + base64.RawURLEncoding.EncodeToString(sum[:])
+	case Base64Fingerprint:
+		return prefix + base64.StdEncoding.EncodeToString(sum[:])
+	case Base64UrlFingerprint:
+		return prefix + base64.URLEncoding.EncodeToString(sum[:])
+	case HexFingerprint:
+		return prefix + strings.ToLower(hex.EncodeToString(sum[:]))
+	case EmojiFingerprint:
+		return prefix + emoji.Emoji(sum[:])
+	default:
+		return ""
+	}
+}
+
+// FormatFingerprint gets a public key from an authorized_keys file used in
+// OpenSSH and returns the fingerprint in the following format:
+//
+//	<size> SHA256:<base64-raw-fingerprint> <comment> (<type)
```

Comment:
```suggestion
// FormatFingerprint returns an OpenSSH public key fingerprint in the following format:
//
//	<size> SHA256:<base64-raw-fingerprint> <comment> (<type)
```

Reply:
Fixed with 613cf63, but I've haven't used your suggestion.

Reply:
Yeah, I'm OK with the new one. At first it read like the code would also read `authorized_keys`, which it obviously didn't. Now it makes sense, as it's more of a description of what type of public key it is.